File: src/test/kotlin/xtdb/aws/MinioTest.kt
package xtdb.aws

import clojure.java.api.Clojure
import io.minio.MakeBucketArgs
import io.minio.MinioClient
import org.jetbrains.exposed.sql.Database.Companion.connect
import org.jetbrains.exposed.sql.transactions.transaction
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import org.testcontainers.containers.MinIOContainer
import software.amazon.awssdk.regions.Region.AWS_ISO_GLOBAL
import xtdb.api.Xtdb
import xtdb.api.log.Log
import xtdb.api.storage.Storage
import xtdb.cache.DiskCache
import xtdb.database.Database
import xtdb.symbol
import xtdb.util.asPath
import java.nio.file.Path
import java.util.*
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.time.Duration.Companion.seconds
import kotlin.time.toJavaDuration

@Tag("integration")
class MinioTest : S3Test() {

    companion object {
        private var wasRunning = false
        private val container = MinIOContainer("minio/minio")

        @JvmStatic
        @BeforeAll
        fun setUpMinio() {
            if (container.isRunning) wasRunning = true else container.start()

            MinioClient.builder()
                .endpoint(container.s3URL).credentials(container.userName, container.password)
                .build()
                .apply { makeBucket(MakeBucketArgs.builder().bucket(bucket).build()) }
        }

        @JvmStatic
        @AfterAll
        fun tearDownMinio() {
            if (!wasRunning) container.stop()
        }
    }

    override fun openObjectStore(prefix: Path) = S3.s3(bucket) {
        endpoint(container.s3URL)
        credentials(container.userName, container.password)
        region(AWS_ISO_GLOBAL)
        prefix(prefix)
        pathStyleAccessEnabled(true)
    }.openObjectStore()

    @Test
    fun writeBlock(@TempDir nodeDir: Path) {
        Clojure.`var`("clojure.core/require").invoke("xtdb.types".symbol)

        val xtdbLog = Log.localLog(nodeDir.resolve("xt-log"))
        val xtdbStorage = Storage.local(nodeDir.resolve("xt-objs"))

        Xtdb.openNode {
            diskCache(DiskCache.factory(nodeDir.resolve("disk-cache")))

            log(xtdbLog)
            storage(xtdbStorage)
        }.use { node ->
            node.connection.use { conn ->
                conn.createStatement().use { stmt ->
                    stmt.execute("""ATTACH DATABASE foo WITH $$
                        log: !Local { path: "${nodeDir.resolve("foo-log")}" }
                        storage: !Remote 
                          objectStore: !S3 
                            bucket: "$bucket"
                            endpoint: "${container.s3URL}"
                            region: "${AWS_ISO_GLOBAL.id()}"
                            credentials:
                              accessKey: "${container.userName}"
                              secretKey: "${container.password}"
                            pathStyleAccessEnabled: true
                            prefix: "${UUID.randomUUID()}"
                        $$"""
                    )
                }
            }

            transaction(db = connect({ node.createConnectionBuilder().database("foo").build() })) {
                exec("INSERT INTO foo RECORDS {_id: 1}")
            }

            (node as Xtdb.XtdbInternal).dbCatalog.let { cat ->
                cat.primary.sendFlushBlockMessage()
                cat["foo"]!!.sendFlushBlockMessage()
                cat.syncAll(2.seconds.toJavaDuration())
            }

            transaction(db = connect(node)) {
                exec("SELECT * FROM foo.foo") { rs ->
                    assertTrue(rs.next())
                    assertEquals(1, rs.getInt(1))
                    assertFalse(rs.next())
                }
            }
        }

        Xtdb.openNode {
            diskCache(DiskCache.factory(nodeDir.resolve("disk-cache2")))
            log(xtdbLog)
            storage(xtdbStorage)
        }.use { node ->
            val dbCatalog = (node as Xtdb.XtdbInternal).dbCatalog

            assertEquals(setOf("xtdb", "foo"), dbCatalog.databaseNames.toSet())

            transaction(db = connect(node)) {
                exec("SELECT * FROM foo.foo") { rs ->
                    assertTrue(rs.next())
                    assertEquals(1, rs.getInt(1))
                    assertFalse(rs.next())
                }
            }
        }
    }
}

File: src/test/clojure/xtdb/aws/minio_test.clj
(ns xtdb.aws.minio-test
  (:require [clojure.test :as t]
            [xtdb.api :as xt]
            [xtdb.buffer-pool-test :as bp-test]
            [xtdb.node :as xtn]
            [xtdb.object-store :as os]
            [xtdb.object-store-test :as os-test]
            [xtdb.test-util :as tu]
            [xtdb.util :as util])
  (:import [java.nio ByteBuffer]
           [java.nio.file Path]
           [software.amazon.awssdk.services.s3 S3AsyncClient]
           [software.amazon.awssdk.services.s3.model ListMultipartUploadsRequest ListMultipartUploadsResponse]
           [xtdb.api.storage ObjectStore]
           [xtdb.aws S3]
           [xtdb.buffer_pool RemoteBufferPool]
           [xtdb.multipart IMultipartUpload SupportsMultipart]))

;; To run these, run the MinIO and Kafka containers in the docker-compose file
;; http://localhost:9001, minioadmin/minioadmin
;; Identity -> Users -> Create: xtdb/test-password, policy: readwrite
;; Bucket: xtdb, policy: readwrite

(def ^:const bucket "xtdb")

(def test-creds
  {:access-key "xtdb"
   :secret-key "test-password"})

(defn object-store ^xtdb.api.storage.ObjectStore [prefix]
  (let [{:keys [access-key secret-key]} test-creds]
    (-> (S3/s3 bucket)
        (.prefix (util/->path (str prefix)))
        (.credentials access-key secret-key)
        (.endpoint "http://127.0.0.1:9000")
        (.openObjectStore))))

(t/deftest ^:minio put-delete-test
  (with-open [os (object-store (random-uuid))]
    (os-test/test-put-delete os)))

(defn start-node [^Path node-dir prefix]
  (xtn/start-node
   {:storage [:remote {:object-store [:s3 {:bucket bucket
                                           :prefix (str prefix)
                                           :credentials test-creds
                                           :endpoint "http://127.0.0.1:9000"}]
                       :local-disk-cache (.resolve node-dir "local-cache")}]
    :log [:kafka {:topic (str "xtdb.kafka-test." prefix)
                  :bootstrap-servers "localhost:9092"}]}))

(t/deftest ^:minio list-test
  (util/with-tmp-dirs #{local-disk-cache}
    (util/with-open [node (start-node local-disk-cache (random-uuid))]
      (let [buffer-pool (bp-test/fetch-buffer-pool-from-node node)]
        (bp-test/test-list-objects buffer-pool)))))

(t/deftest ^:minio list-test-with-prior-objects
  (util/with-tmp-dirs #{local-disk-cache}
    (let [prefix (random-uuid)]
      (util/with-open [node (start-node local-disk-cache prefix)]
        (let [^RemoteBufferPool buffer-pool (bp-test/fetch-buffer-pool-from-node node)]
          (bp-test/put-edn buffer-pool (util/->path "alice") :alice)
          (bp-test/put-edn buffer-pool (util/->path "alan") :alan)
          (Thread/sleep 100)
          (t/is (= [(os/->StoredObject "alan" 5) (os/->StoredObject "alice" 6)]
                   (.listAllObjects buffer-pool)))))

      (util/with-open [node (start-node local-disk-cache prefix)]
        (let [^RemoteBufferPool buffer-pool (bp-test/fetch-buffer-pool-from-node node)]
          (t/testing "prior objects will still be there, should be available on a list request"
          (Thread/sleep 100)
            (t/is (= [(os/->StoredObject "alan" 5) (os/->StoredObject "alice" 6)]
                     (.listAllObjects buffer-pool))))

          (t/testing "should be able to add new objects and have that reflected in list objects output"
            (bp-test/put-edn buffer-pool (util/->path "alex") :alex)
            (Thread/sleep 100)
            (t/is (= [(os/->StoredObject "alan" 5) (os/->StoredObject "alex" 5) (os/->StoredObject "alice" 6)]
                     (.listAllObjects buffer-pool)))))))))

(t/deftest ^:minio multiple-node-list-test
  (util/with-tmp-dirs #{local-disk-cache}
    (let [prefix (random-uuid)]
      (util/with-open [node-1 (start-node local-disk-cache prefix)
                       node-2 (start-node local-disk-cache prefix)]
        (let [^RemoteBufferPool buffer-pool-1 (bp-test/fetch-buffer-pool-from-node node-1)
              ^RemoteBufferPool buffer-pool-2 (bp-test/fetch-buffer-pool-from-node node-2)]
          (bp-test/put-edn buffer-pool-1 (util/->path "alice") :alice)
          (bp-test/put-edn buffer-pool-2 (util/->path "alan") :alan)
          (Thread/sleep 1000)
          (t/is (= [(os/->StoredObject "alan" 5) (os/->StoredObject "alice" 6)]
                   (.listAllObjects buffer-pool-1)))

          (t/is (= [(os/->StoredObject "alan" 5) (os/->StoredObject "alice" 6)]
                   (.listAllObjects buffer-pool-2))))))))

(t/deftest ^:minio multipart-start-and-cancel
  (with-open [os (object-store (random-uuid))]
    (let [multipart-key (util/->path "test-multi-created")
          multipart-upload ^IMultipartUpload  @(.startMultipart ^SupportsMultipart os multipart-key)]

      ;; MinIO doesn't seem to support listing incomplete uploads

      (t/testing "Call to abort a multipart upload should work"
        (t/is @(.abort multipart-upload))))))

(t/deftest ^:minio multipart-put-test
  (with-open [os (object-store (random-uuid))]
    (let [prefix (:prefix os)
          multipart-upload ^IMultipartUpload @(.startMultipart ^SupportsMultipart os (util/->path "test-multi-put"))
          part-size (* 5 1024 1024)
          file-part-1 (os-test/generate-random-byte-buffer part-size)
          file-part-2 (os-test/generate-random-byte-buffer part-size)]

      ;; Uploading parts to multipart upload
      @(.uploadPart multipart-upload file-part-1)
      @(.uploadPart multipart-upload file-part-2)

      (t/testing "Call to complete a multipart upload should work - should be removed from the upload list"
        @(.complete multipart-upload)
        (let [list-multipart-uploads-response @(.listMultipartUploads ^S3AsyncClient (:client os)
                                                                      (-> (ListMultipartUploadsRequest/builder)
                                                                          (.bucket bucket)
                                                                          (.prefix (str prefix))
                                                                          ^ListMultipartUploadsRequest (.build)))
              uploads (.uploads ^ListMultipartUploadsResponse list-multipart-uploads-response)]
          (t/is (= [] uploads) "uploads should be empty")))

      (t/testing "Multipart upload works correctly - file present and contents correct"
        (t/is (= [(os/->StoredObject (util/->path "test-multi-put")
                                     (* 2 part-size))]
                 (.listAllObjects ^ObjectStore os)))

        (let [^ByteBuffer uploaded-buffer @(.getObject ^ObjectStore os (util/->path "test-multi-put"))]
          (t/testing "capacity should be equal to total of 2 parts"
            (t/is (= (* 2 part-size) (.capacity uploaded-buffer)))))))))

(t/deftest ^:minio node-level-test
  (util/with-tmp-dirs #{local-disk-cache}
    (util/with-open [node (start-node local-disk-cache (random-uuid))]
      (let [^RemoteBufferPool buffer-pool (bp-test/fetch-buffer-pool-from-node node)]
        ;; Submit some documents to the node
        (t/is (= true
                 (:committed? (xt/execute-tx node [[:put-docs :bar {:xt/id "bar1"}]
                                                   [:put-docs :bar {:xt/id "bar2"}]
                                                   [:put-docs :bar {:xt/id "bar3"}]]))))
  
        ;; Ensure finish-chunk! works
        (t/is (nil? (tu/finish-chunk! node)))
  
        ;; Ensure can query back out results
        (t/is (= [{:e "bar2"} {:e "bar1"} {:e "bar3"}]
                 (xt/q node '(from :bar [{:xt/id e}]))))
  
        ;; Ensure some files written to buffer-pool
        (t/is (seq (.listAllObjects buffer-pool)))))))

;; Using large enough TPCH ensures multiparts get properly used within the bufferpool
(comment
  (import [java.time Duration])
  (require '[xtdb.datasets.tpch :as tpch]
           '[integrant.core :as ig])

  (t/deftest ^:minio tpch-test-node
    (util/with-tmp-dirs #{local-disk-cache}
                        (util/with-open [node (start-node local-disk-cache (str (random-uuid)))]
                                        ;; Submit tpch docs
                                        (-> (tpch/submit-docs! node 0.05)
            (tu/then-await-tx node (Duration/ofHours 1)))

                                        ;; Ensure finish-chunk! works
                                        (t/is (nil? (tu/finish-chunk! node)))

                                        (let [{:keys [^ObjectStore object-store] :as buffer-pool} (val (first (ig/find-derived (:system node) :xtdb/storage)))]
                                          (t/is (instance? RemoteBufferPool buffer-pool))
                                          (t/is (instance? ObjectStore object-store))
                                          ;; Ensure some files are written
                                          (t/is (seq (.listAllObjects object-store))))))))
